dataService.js

```diff
@@ -0,0 +1,94 @@
+// Maps
+let productMap;
+let categoryMap;
+let combinedProductMap;
+
+/*
+ * Method to create and return map of products if its not already created
+ */
+function getProducts() {
+
+  if(productMap) {
+    return productMap;
+  }
+
+  const productData = require('./src/data/products.json');
+
+  // Method variable to be stored in productMap
+  function createProductMap() {
+    const productMapToBeReturned = {};
+
+    productData.products.forEach(product => {
+      productMapToBeReturned[product.id] = product;
+    });
+    return productMapToBeReturned;
+  }
+
+  return productMap = createProductMap();
+}
+
+/*
+ * Method to create and return map of categories if not already created
+ */
+function getCategories() {
+
+  if (categoryMap) {
+    return categoryMap;
+  }
+
+  const categoryData = require('./src/data/categories.json');
+
+  // Method variable to be stored in categoryMap
+  function createCategoryMap() {
+    const categoryMapToBeReturned = {};
+
+    categoryData.categories.forEach(category => {
+      categoryMapToBeReturned[category.id] = category.categoryName;
+    });
+    return categoryMapToBeReturned;
+  }
+
+  return categoryMap = createCategoryMap();
+}
+
+/*
+ * Method to combine product and category data into combinedProductMap
+ */
+function combineProductsWithCategories() {
+
+  // Creating categoryMap and getting required data
+  getCategories();
+  const productData = require('./src/data/products.json');
+
+  combinedProductMap = [];
+
+  productData.products.forEach(product => {
+    // Making deep copy to prevent changes in actual data
+    let productToBeReturned = Object.assign({}, product);
```

Comment:
Here, We actually simply clone the object with new reference. Is it enough?
```suggestion
    let productToBeReturned = { ...product };
```

Reply:
I tested that and its causing changes in JSON structure, making things more complicated and difficult, we better off using `Object.assign({}, product)`

Reply:
Why not just simply 
```suggestion
    let productToBeReturned = product;
```

Reply:
Doing that would cause change in actual data, which is undesirable

Reply:
Done!